File: src/app/zap-templates/common/simulated-clusters/clusters/DiscoveryCommands.js
/*
 *
 *    Copyright (c) 2021 Project CHIP Authors
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

/*
 * This file declares test suite utility methods for discovery.
 *
 * Each method declared in this file needs to be implemented on a per-language
 * basis and allows exposing  methods to the test suites that are not part
 * of the regular cluster set of APIs.
 *
 */

const kTypeArgument = {
  name : 'type',
  type : 'CHAR_STRING',
};

const kNumberValueArgument = {
  name : 'value',
  type : 'INT64U',
};

const kStringValueArgument = {
  name : 'value',
  type : 'CHAR_STRING',
};

const kDefaultResponse = {
  arguments : [
    { name : 'hostName', type : 'CHAR_STRING', chipType : 'chip::CharSpan' }, //
    { name : 'instanceName', type : 'CHAR_STRING', chipType : 'chip::CharSpan' }, //
    { name : 'longDiscriminator', type : 'INT16U', chipType : 'uint16_t' }, //
    { name : 'shortDiscriminator', type : 'INT8U', chipType : 'uint8_t' }, //
    { name : 'vendorId', type : 'INT16U', chipType : 'uint16_t' }, //
    { name : 'productId', type : 'INT16U', chipType : 'uint16_t' }, //
    { name : 'commissioningMode', type : 'INT8U', chipType : 'uint8_t' }, //
    { name : 'deviceType', type : 'INT16U', chipType : 'uint16_t' }, //
    { name : 'deviceName', type : 'CHAR_STRING' }, //
    { name : 'rotatingId', type : 'OCTET_STRING', chipType : 'chip::ByteSpan' }, //
    { name : 'rotatingIdLen', type : 'INT64U', chipType : 'uint64_t' }, //
    { name : 'pairingHint', type : 'INT16U', chipType : 'uint16_t' }, //
    { name : 'pairingInstruction', type : 'CHAR_STRING' }, //
    { name : 'supportsTcp', type : 'BOOLEAN', chipType : 'bool' }, //
    { name : 'numIPs', type : 'INT8U', chipType : 'uint8_t' }, //
    { name : 'port', type : 'INT16U', chipType : 'uint16_t' }, //
    { name : 'mrpRetryIntervalIdle', type : 'INT32U', chipType : 'uint32_t', isOptional : true }, //
    { name : 'mrpRetryIntervalActive', type : 'INT32U', chipType : 'uint32_t', isOptional : true }, //
  ]
};

//
// Commissionable
//

const FindCommissionable = {
  name : 'FindCommissionable',
  arguments : [],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionableByShortDiscriminator = {
  name : 'FindCommissionableByShortDiscriminator',
  arguments : [ kNumberValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionableByLongDiscriminator = {
  name : 'FindCommissionableByLongDiscriminator',
  arguments : [ kNumberValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionableByCompressedFabricId = {
  name : 'FindOperationalByCompressedFabricId',
  arguments : [ kNumberValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionableByCommissioningMode = {
  name : 'FindCommissionableByCommissioningMode',
  arguments : [],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionableByVendorId = {
  name : 'FindCommissionableByVendorId',
  arguments : [ kNumberValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionableByDeviceType = {
  name : 'FindCommissionableByDeviceType',
  arguments : [ kNumberValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionableByName = {
  name : 'FindCommissionableByName',
  arguments : [ kStringValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

//
// Commissioner
//

const FindCommissioner = {
  name : 'FindCommissioner',
  arguments : [],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionerByVendorId = {
  name : 'FindCommissionableByVendorId',
  arguments : [ kNumberValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const FindCommissionerByDeviceType = {
  name : 'FindCommissionerByDeviceType',
  arguments : [ kNumberValueArgument ],
  hasSpecificResponse : true,
  responseName : 'DiscoveryCommandResponse',
  response : kDefaultResponse
};

const commands = [
  // Commissionable
  FindCommissionable,
  FindCommissionableByShortDiscriminator,
  FindCommissionableByLongDiscriminator,
  FindCommissionableByCommissioningMode,
  FindCommissionableByVendorId,
  FindCommissionableByDeviceType,
  FindCommissionableByName,
  // Commissioner
  FindCommissioner,
  FindCommissionerByVendorId,
  FindCommissionerByDeviceType,
];

const DiscoveryCommands = {
  name : 'DiscoveryCommands',
  commands : commands,
};

//
// Module exports
//
exports.cluster = DiscoveryCommands;

File: scripts/py_matter_yamltests/matter/yamltests/pseudo_clusters/clusters/discovery_commands.py
#
#    Copyright (c) 2023 Project CHIP Authors
#
#    Licensed under the Apache License, Version 2.0 (the 'License');
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an 'AS IS' BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.

from ..pseudo_cluster import PseudoCluster

_DEFINITION = '''<?xml version="1.0"?>
<configurator>
<cluster>
    <name>DiscoveryCommands</name>
    <code>0xFFF1FD05</code>

    <command source="client" code="0" name="FindCommissionable" response="FindResponse">
    </command>

    <command source="client" code="1" name="FindCommissionableByShortDiscriminator" response="FindResponse">
        <arg name="value" type="int16u"/>
    </command>

    <command source="client" code="2" name="FindCommissionableByLongDiscriminator" response="FindResponse">
        <arg name="value" type="int16u"/>
    </command>

    <command source="client" code="3" name="FindCommissionableByCommissioningMode" response="FindResponse">
    </command>

    <command source="client" code="4" name="FindCommissionableByVendorId" response="FindResponse">
        <arg name="value" type="vendor_id"/>
    </command>

    <command source="client" code="5" name="FindCommissionableByDeviceType" response="FindResponse">
        <arg name="value" type="devtype_id"/>
    </command>

    <command source="client" code="6" name="FindCommissioner" response="FindCommissionerResponse">
    </command>

    <command source="client" code="7" name="FindCommissionerByVendorId" response="FindCommissionerResponse">
        <arg name="value" type="vendor_id"/>
    </command>

    <command source="client" code="8" name="FindCommissionerByDeviceType" response="FindCommissionerResponse">
        <arg name="value" type="devtype_id"/>
    </command>

    <command source="server" code="9" name="FindResponse">
        <arg name="hostName" type="char_string"/>
        <arg name="instanceName" type="char_string"/>
        <arg name="longDiscriminator" type="int16u"/>
        <arg name="shortDiscriminator" type="int16u"/>
        <arg name="vendorId" type="vendor_id"/>
        <arg name="productId" type="int16u"/>
        <arg name="commissioningMode" type="int8u"/>
        <arg name="deviceType" type="devtype_id"/>
        <arg name="deviceName" type="char_string"/>
        <arg name="rotatingId" type="octet_string"/>
        <arg name="rotatingIdLen" type="int64u"/>
        <arg name="pairingHint" type="int16u"/>
        <arg name="pairingInstruction" type="char_string"/>
        <arg name="supportsTcpClient" type="boolean"/>
        <arg name="supportsTcpServer" type="boolean"/>
        <arg name="numIPs" type="int8u"/>
        <arg name="port" type="int16u"/>
        <arg name="mrpRetryIntervalIdle" type="int32u" optional="true"/>
        <arg name="mrpRetryIntervalActive" type="int32u" optional="true"/>
        <arg name="mrpRetryActiveThreshold" type="int16u" optional="true"/>
        <arg name="isICDOperatingAsLIT" type="boolean" optional="true"/>
    </command>
</cluster>
</configurator>
'''


class DiscoveryCommands(PseudoCluster):
    name = 'DiscoveryCommands'
    definition = _DEFINITION
